scripts/tybalt_vae.py

```diff
@@ -0,0 +1,320 @@
+# 
+# The VAE is compiled with an Adam optimizer and built-in custom loss function. The `loss_weights` parameter ensures beta is updated at each epoch end callback
+
+# In[12]:
+
+adam = optimizers.Adam(lr=learning_rate)
+vae_layer = CustomVariationalLayer()([rnaseq_input, rnaseq_reconstruct])
+vae = Model(rnaseq_input, vae_layer)
+vae.compile(optimizer=adam, loss=None, loss_weights=[beta])
+
+vae.summary()
+
+
+# In[13]:
+
+# Visualize the connections of the custom VAE model
+output_model_file = os.path.join('figures', 'onehidden_vae_architecture.png')
+plot_model(vae, to_file=output_model_file)
+
+SVG(model_to_dot(vae).create(prog='dot', format='svg'))
+
+
+# ## Train the model
+# 
+# The training data is shuffled after every epoch and 10% of the data is heldout for calculating validation loss.
+
+# In[14]:
+
+get_ipython().run_cell_magic('time', '', 'hist = vae.fit(np.array(rnaseq_train_df),\n               shuffle=True,\n               epochs=epochs,\n               batch_size=batch_size,\n               validation_data=(np.array(rnaseq_test_df), np.array(rnaseq_test_df)),\n               callbacks=[WarmUpCallback(beta, kappa)])')
+
+
+# In[15]:
+
+# Visualize training performance
+history_df = pd.DataFrame(hist.history)
+hist_plot_file = os.path.join('figures', 'onehidden_vae_training.pdf')
+ax = history_df.plot()
+ax.set_xlabel('Epochs')
+ax.set_ylabel('VAE Loss')
+fig = ax.get_figure()
+fig.savefig(hist_plot_file)
+
+
+# ## Compile and output trained models
+# 
+# We are interested in:
+# 
+# 1. The model to encode/compress the input gene expression data
+#   * Can be possibly used to compress other tumors
+# 2. The model to decode/decompress the latent space back into gene expression space
+#   * This is our generative model
+# 3. The latent space compression of all pan cancer TCGA samples
+#   * Non-linear reduced dimension representation of tumors can be used as features for various tasks
+#     * Supervised learning tasks predicting specific gene inactivation events
+#     * Interpolating across this space to observe how gene expression changes between two cancer states
+# 4. The weights used to compress each latent node
+#   * Potentially indicate learned biology differentially activating tumors
+
+# ### Encoder model
+
+# In[16]:
+
+# Model to compress input
+encoder = Model(rnaseq_input, z_mean_encoded)
+
+
+# In[17]:
+
+# Encode rnaseq into the hidden/latent representation - and save output
+encoded_rnaseq_df = encoder.predict_on_batch(rnaseq_df)
+encoded_rnaseq_df = pd.DataFrame(encoded_rnaseq_df, index=rnaseq_df.index)
+
+encoded_rnaseq_df.columns.name = 'sample_id'
+encoded_rnaseq_df.columns = encoded_rnaseq_df.columns + 1
+encoded_file = os.path.join('data', 'encoded_rnaseq_onehidden_warmup_batchnorm.tsv')
+encoded_rnaseq_df.to_csv(encoded_file, sep='\t')
+
+
+# In[18]:
+
+# What are the most and least activated nodes
+top_active_nodes = encoded_rnaseq_df.sum(axis=0).sort_values(ascending=False)
```

Comment:
I used this here when I was first troubleshooting the model - often times I would get several latent features that summed to zero or other very small activities. For some reason certain parameter combinations dropped the weights to zero and were basically optimizing bias vectors. So the loss would still go down, but it wouldn't actually learn anything.

To summarize:

> why is summing the activities the most informative?

It's not really informative of anything other than that the latent features are learning some sort of distribution of activities.

Perhaps it be helpful if I add something similar as a comment